core/src/main/scala/org/apache/spark/api/python/PythonRunner.scala

```diff
@@ -180,6 +180,7 @@ private[spark] abstract class BasePythonRunner[IN, OUT](
     def shutdownOnTaskCompletion() {
       assert(context.isCompleted)
       this.interrupt()
+      this.join()
```

Comment:
Ur..I think this change just hangs the execution.

Reply:
This just blocks the current Thread and wait for interrupt to take effect?

Reply:
It causes a deadlock on the lock on task context.

Reply:
Hi, @viirya. Can you explain more about 'deadlock' please ? I just notice that it's possible we fall into deadlock if WritedThread trying to acquire lock on TaskContext after `this.join()` invoked, but I haven't found where it did happens.

Reply:
For example, when the writer thread is reading from a cached relation, it needs to add task complete listener. At the moment, it needs to acquire the lock on task context.

Reply:
hmm...`this.join()` is called within task runner thread while holding the lock on task context. And once `join()` is called, WriteThread would throw `InterruptedException`, then, we run into `case _: InterruptedException =>` code branch within WriteThread. In this branch, if we acquire the lock on task context, we'll fall into deadlock, otherwise, WriteThread exit normally. Right ?

And if WriteThread wants to acquire the lock on task context(as you mentioned, adding task complete listener due to read from a cached relation) out side of above code branch (or say, within its normal execution process), it just preform a normal lock competition between WriteThread and task runner thread. WDYT ?

Please correct me if I mis-understanding on some points.